sys/net/gnrc/sock/ip/gnrc_sock_ip.c

```diff
@@ -0,0 +1,199 @@
+{
+    gnrc_pktsnip_t *pkt;
+    sock_ip_ep_t tmp;
+    int res;
+
+    assert((sock != NULL) && (data != NULL) && (max_len > 0));
+    if (sock->local.family == 0) {
+        return -EADDRNOTAVAIL;
+    }
+    tmp.family = sock->local.family;
+    res = gnrc_sock_recv((gnrc_sock_reg_t *)sock, &pkt, timeout, &tmp);
+    if (res < 0) {
+        return res;
+    }
+    if (pkt->size > max_len) {
+        gnrc_pktbuf_release(pkt);
+        return -ENOBUFS;
+    }
+    if (remote != NULL) {
+        /* return remote to possibly block if wrong remote */
+        memcpy(remote, &tmp, sizeof(tmp));
+    }
+    if ((sock->remote.family != AF_UNSPEC) &&   /* check remote end-point if set */
+        /* We only have IPv6 for now, so just comparing the whole end point
+         * should suffice */
+        ((memcmp(&sock->remote.addr, &ipv6_addr_unspecified,
+                 sizeof(ipv6_addr_t)) != 0) &&
+         (memcmp(&sock->remote.addr, &tmp.addr, sizeof(ipv6_addr_t)) != 0))) {
+        gnrc_pktbuf_release(pkt);
+        return -EPROTO;
+    }
+    memcpy(data, pkt->data, pkt->size);
+    gnrc_pktbuf_release(pkt);
+    return (int)pkt->size;
+}
+
+ssize_t sock_ip_send(sock_ip_t *sock, const void *data, size_t len,
+                     uint8_t proto, const sock_ip_ep_t *remote)
+{
+    int res;
+    gnrc_pktsnip_t *pkt;
+    sock_ip_ep_t local;
+    sock_ip_ep_t rem;
+
+    assert((sock != NULL) || (remote != NULL));
+    assert((len == 0) || (data != NULL)); /* (len != 0) => (data != NULL) */
+    if ((remote != NULL) && (sock != NULL) &&
+        (sock->local.netif != SOCK_ADDR_ANY_NETIF) &&
+        (remote->netif != SOCK_ADDR_ANY_NETIF) &&
+        (sock->local.netif != remote->netif)) {
+        return -EINVAL;
+    }
+    if ((remote == NULL) &&
+        /* sock can't be NULL as per assertion above */
+        (sock->remote.family == AF_UNSPEC)) {
+        return -ENOTCONN;
+    }
+    else if ((remote != NULL) && (gnrc_ep_addr_any(remote))) {
+        return -EINVAL;
+    }
+    /* compiler evaluates lazily so this isn't a redundundant check and cppcheck
+     * is being weird here anyways */
+    /* cppcheck-suppress nullPointerRedundantCheck */
+    /* cppcheck-suppress nullPointer */
+    if ((sock == NULL) || (sock->local.family == AF_UNSPEC)) {
+        /* no sock or sock currently unbound */
+        memset(&local, 0, sizeof(local));
+    }
+    else {
+        if (sock != NULL) {
+            proto = (uint8_t)sock->reg.entry.demux_ctx;
+        }
+        memcpy(&local, &sock->local, sizeof(local));
+    }
+    if (remote == NULL) {
+        /* sock can't be NULL at this point */
+        memcpy(&rem, &sock->remote, sizeof(rem));
+    }
+    else {
+        memcpy(&rem, remote, sizeof(rem));
+    }
+    if ((remote != NULL) && (remote->family == AF_UNSPEC) &&
+        (sock->remote.family != AF_UNSPEC)) {
+        /* remote was set on create so take its family */
+        rem.family = sock->remote.family;
+    }
```

Comment:
@miri64 
The assert above indicates that `remote` or `sock` must be non-NULL.
At this if statement we check that `remote` is not NULL. But in that case it is possible that `sock` is NULL. We don't check that before using `sock`.

Reply:
Confirmed and fixed in https://github.com/RIOT-OS/RIOT/pull/14963